content/guides/dapps/cash-app.md

```diff
@@ -0,0 +1,2505 @@
+
+        require!(*cash_account.owner == ctx.accounts.signer.key(), ErrorCode::InvalidSigner);
+
+        **from_cash_account.try_borrow_mut_lamports()? -= amount;
+        **to_cash_account.try_borrow_mut_lamports()? += amount;
+
+        Ok(())
+    }
+}
+
+#[derive(Accounts)]
+#[instruction(recipient: Pubkey)]
+pub struct TransferFunds<'info> {
+    #[account(
+        mut,
+        seeds = [b"cash-account", signer.key().as_ref()],
+        bump,
+    )]
+    pub from_cash_account: Account<'info, CashAccount>,
+
+    #[account(
+        mut,
+        seeds = [b"cash-account", recipient.key().as_ref()],
+        bump,
+    )]
+    pub to_cash_account: Account<'info, CashAccount>,
+    pub system_program: Program<'info, System>,
+    pub signer: Signer<'info>,
+}
+```
+
+In the above instruction, the `TransferFunds` Context data structure consists of
+an additional account. The `Context` is a struct that includes references to all
+the accounts needed for the operation. Since we need information from both the
+sender and recipient accounts for this instruction, we need to include both
+accounts in the `Context`.
+
+We are once again directly transferring lamports between accounts, since the
+program owns the `cash_account` account. The seeds for the cash account PDAs are
+created from the public key of the cash account owner so the instruction needs
+to take the recipient's public key as a parameter and pass that to the
+`TransferFunds` Context data structure. Then the `cash_account` PDA can be
+derived for both the `from_cash_account` and the `to_cash_account`.
+
+Since both of the accounts are listed in the `#[derive(Accounts)]` macro, they
+are deserialized and validated so you can simply call both of the accounts with
+the Context `ctx` to get the account info and update the account balances from
+there.
+
+To be able to send funds to another user, similar to Cash App, both users must
+have created an account. We're sending funds to the user's `cash_account` PDA,
+not the user's wallet. So each user needs to initialize a cash account by
+calling the `initialize_account` instruction to create their unique PDA derived
+from their wallet public key. We'll need to keep this in mind when designing the
+UI/UX of the onboarding process for this dApp later on to ensure every user
+calls the `initialize_account` instruction when signing up for an account.
+
+Now that the basic payment functionality is enabled, we want to be able to
+interact with friends. So we need to add instructions for adding friends,
+requesting payments from friends, and accepting/rejecting payment requests.
+
+Adding a friend is as simple as just pushing a new public key to the `friends`
+vector in the `CashAccount` state.
+
+```rust
+#[program]
+pub mod cash_app {
+    use super::*;
+
+    ...
+    pub fn add_friend(ctx: Context<AddFriend>, pubkey: Pubkey) -> Result<()> {
+        let cash_account = &mut ctx.accounts.cash_account;
+        cash_account.friends.push(pubkey);
```

Comment:
It would be good to note the limitations of this design. Vec can only grow to so many friends.